File: backend/src/libs/firebase.js
const admin = require('firebase-admin');
const fs = require('fs');
const path = require('path');

const DEFAULT_CREDENTIALS_FILE = 'pi-3-286ed-firebase-adminsdk-fbsvc-9edefa08b0.json';
const BACKEND_ROOT = path.join(__dirname, '..', '..');

let initialized = false;

function resolveCredentialsPath(configured) {
  const candidates = [];

  if (configured && String(configured).trim()) {
    const value = String(configured).trim();
    candidates.push(path.isAbsolute(value) ? value : path.join(BACKEND_ROOT, value));
  }

  candidates.push(path.join(BACKEND_ROOT, DEFAULT_CREDENTIALS_FILE));

  for (const filePath of candidates) {
    if (fs.existsSync(filePath)) return filePath;
  }

  return null;
}

function initFirebaseAdmin() {
  if (initialized) return;

  const storageBucket =
    process.env.FIREBASE_STORAGE_BUCKET || 'pi-3-286ed.firebasestorage.app';

  const credentialsPath = resolveCredentialsPath(process.env.FIREBASE_CREDENTIALS_FILE);
  if (!credentialsPath) {
    console.warn(
      `[backend] Arquivo de credenciais Firebase nao encontrado. Defina FIREBASE_CREDENTIALS_FILE ou coloque ${DEFAULT_CREDENTIALS_FILE} na pasta backend/.`
    );
    return;
  }

  const raw = fs.readFileSync(credentialsPath, 'utf8');
  const serviceAccount = JSON.parse(raw);

  const options = {
    credential: admin.credential.cert(serviceAccount),
    projectId: serviceAccount.project_id
  };
  if (storageBucket && storageBucket.trim().length) {
    options.storageBucket = storageBucket.trim();
  }

  if (admin.apps.length === 0) {
    admin.initializeApp(options);
  }

  initialized = true;
  console.log(`[backend] Firebase Admin inicializado (${path.basename(credentialsPath)})`);
}

function getAdmin() {
  initFirebaseAdmin();
  return admin;
}

module.exports = { initFirebaseAdmin, getAdmin, DEFAULT_CREDENTIALS_FILE };

File: backend/config/initializers/firebase.rb
credentials_path = Rails.root.join("config/firebase/credentials.json")

if credentials_path.exist?
  raw = JSON.parse(credentials_path.read)
  Rails.application.config.firebase = {
    credentials_path: credentials_path.to_s,
    project_id: raw["project_id"],
    storage_bucket: ENV.fetch("FIREBASE_STORAGE_BUCKET", "#{raw['project_id']}.firebasestorage.app"),
    credentials: raw
  }.freeze
else
  Rails.logger.warn("[Firebase] Credenciais nao encontradas em #{credentials_path}")
  Rails.application.config.firebase = {
    credentials_path: nil,
    project_id: ENV["FIREBASE_PROJECT_ID"],
    storage_bucket: ENV["FIREBASE_STORAGE_BUCKET"],
    credentials: nil
  }.freeze
end
